ouroboros-network/src/Ouroboros/Network/Socket.hs

```diff
@@ -62,6 +74,84 @@ import qualified Ouroboros.Network.Server.Socket as Server
 maxTransmissionUnit :: Int64
 maxTransmissionUnit = 4 * 1440
 
+type ConnectionTable = TVar IO (M.Map Socket.SockAddr ConnectionTableEntry)
+
+data ConnectionTableEntry = ConnectionTableEntry {
+      cteRefs           :: ![TVar IO Int]
+    , cteLocalAddresses :: !(S.Set Socket.SockAddr)
+    }
+
+data ConnectionTableRef =
+      ConnectionTableCreate
+    | ConnectionTableExist
+    | ConnectionTableDone
+    deriving Show
+
+newConnectionTable :: IO (ConnectionTable)
+newConnectionTable =  newTVarM M.empty
+
+addConnection
+    :: ConnectionTable
+    -> Socket.SockAddr
+    -> Socket.SockAddr
+    -> [TVar IO Int]
+    -> STM IO ()
+addConnection tblVar remoteAddr localAddr refs =
+    readTVar tblVar >>= M.alterF fn remoteAddr >>= writeTVar tblVar
+  where
+    fn :: Maybe ConnectionTableEntry -> STM IO (Maybe ConnectionTableEntry)
+    fn Nothing = do
+        subRefs refs
+        return $ Just $ (ConnectionTableEntry refs (S.singleton localAddr))
+    fn (Just cte) = do
+          let refs' = refs ++ cteRefs cte
+          subRefs $ refs'
+          return $ Just $ cte {
+                cteRefs = refs'
+              , cteLocalAddresses = S.insert localAddr (cteLocalAddresses cte)
+              }
+
+    subRefs = mapM_ (\a -> modifyTVar' a (\r -> r - 1))
+
+removeConnection
+    :: ConnectionTable
+    -> Socket.SockAddr
+    -> Socket.SockAddr
+    -> IO ()
+removeConnection tblVar remoteAddr localAddr = atomically $
+    readTVar tblVar >>= M.alterF fn remoteAddr >>= writeTVar tblVar
+  where
+    fn :: Maybe ConnectionTableEntry -> STM IO (Maybe ConnectionTableEntry)
+    fn Nothing = return Nothing -- XXX removing non existant address
+    fn (Just ConnectionTableEntry{..}) = do
+        mapM_ (\a -> modifyTVar' a (+ 1)) cteRefs
+        let localAddresses' = S.delete localAddr cteLocalAddresses
+        if null localAddresses'
+            then return Nothing
+            else return $ Just $ ConnectionTableEntry cteRefs localAddresses'
+
+refConnection
+    :: ConnectionTable
+    -> Socket.SockAddr
+    -> TVar IO Int
+    -> IO ConnectionTableRef
+refConnection tblVar remoteAddr refVar = atomically $ do
```

Comment:
Else where `refVar` is called `valencyVar`, could we use `valencyVar` name here as well?